assembly/src/lib.rs

```diff
@@ -1,46 +1,45 @@
 #![no_std]
+#![cfg_attr(all(nightly, not(feature = "std")), feature(error_in_core))]
 
 #[macro_use]
 extern crate alloc;
 
-#[cfg(feature = "std")]
+#[cfg(any(test, feature = "std"))]
 extern crate std;
 
 use vm_core::{
-    code_blocks::CodeBlock,
-    crypto,
+    crypto::hash::RpoDigest,
     errors::KernelError,
     utils::{
         ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable, SliceReader,
     },
-    CodeBlockTable, Felt, Kernel, Operation, Program, StarkField, ONE, ZERO,
+    Felt, Word, ONE, ZERO,
 };
 
-mod library;
-pub use library::{Library, LibraryNamespace, LibraryPath, MaslLibrary, Module, Version};
-
-mod procedures;
-use procedures::{CallSet, NamedProcedure, Procedure};
-pub use procedures::{ProcedureId, ProcedureName};
+use vm_core::{prettier, utils::DisplayHex};
 
+mod assembler;
 pub mod ast;
-use ast::{NAMESPACE_LABEL_PARSER, PROCEDURE_LABEL_PARSER};
-
-mod tokens;
-use tokens::{Token, TokenStream};
-
+mod compile;
+pub mod diagnostics;
 mod errors;
-pub use errors::{AssemblyError, LabelError, LibraryError, ParsingError, PathError};
-
-mod assembler;
-pub use assembler::{Assembler, AssemblyContext};
-
+pub mod library;
+mod parser;
+mod sema;
+#[cfg(any(test, feature = "testing"))]
+pub mod testing;
 #[cfg(test)]
 mod tests;
 
-// RE-EXPORTS
-// ================================================================================================
+pub use self::assembler::{ArtifactKind, Assembler, AssemblyContext};
+pub use self::compile::{Compile, Options as CompileOpts};
+pub use self::errors::AssemblyError;
+pub use self::library::{
+    Library, LibraryError, LibraryNamespace, LibraryPath, MaslLibrary, PathError, Version,
+};
```

Comment:
As a potential counterpoint to that: I think it makes sense to surface the most commonly-used elements of the crate at the top-level, while still allowing access to lesser-used internal components using fully-qualified paths. What you export from the top of the crate doesn't have to be all or nothing, some combination is fine IMO.

That said, we probably don't need to export the error types here, and both `LibraryNamespace` and `LibraryPath` I could go either way on, but `Library`, `MaslLibrary`, and `Version` seem like reasonable top-level exports to me.

The specific exports here were largely based on what was already being exported at the top-level though, I didn't put much effort into rethinking them, since I was looking to minimize breaking changes downstream, but as you noted in the other comment, now is as good a time as any to rethink things like this. Let me know how you'd like to proceed!